user_card_popover: Redesign popover and add accessibility improvements.

sayamsamal opened the pull request:
As part of the popover menu redesign, this PR redesigns the user card popover using the new "popover-menu" tippy theme and improves accessibility by using appropriate ARIA attributes.

Fixes:  #27338

[CZO Discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20user.20card/near/1844285)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

## Screenshots and screen captures:

<details>
<summary>Own User Card</summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/71159abd-b975-4d68-839b-05dc9003d78f"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/9f825864-a659-406f-8bbf-0c564c24cfca"> | 

</p>
</details> 

<details>
<summary>Other's User Card</summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/8de68a41-c4ea-483f-9603-6268a333e449"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/15d97052-f2d2-4eea-bfc7-0affa8f230b4"> | 

</p>
</details> 

<details>
<summary>Other's User Card With Manage Option</summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/35f17b29-acd4-471b-b548-c6ad7f408ca8"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/91a71446-abf5-4471-9881-613b4b5d5ede"> | 

</p>
</details> 

<details>
<summary> Other's User Card From Messages View </summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/d6d4da32-2f80-46b9-baac-e221945f7c3b"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/d54ff304-66c6-4c2e-a2c3-fcb0c80cc088"> | 

</p>
</details> 

<details>
<summary>User Card – Spectator's View </summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/5a2eae1d-31ef-423f-ade7-dd3500101698"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/94a17b47-7045-49a3-99e2-fd5a833f1c41"> | 

</p>
</details> 

<details>
<summary> Deactivated User </summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/221cb15a-cb73-41a1-969a-46cc1ded3413"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/942f3f73-ac9f-4179-be5e-cdc8b3dee4e4"> | 

</p>
</details> 

<details>
<summary>Unknown User Card</summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/9e696b2c-6570-4724-a469-465ceb04f3c8"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/49444182-9b15-452c-b19e-aca86c7dc2e4"> | 

</p>
</details> 

<details>
<summary>Custom Profile field (Text Short + List of options)</summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/2adc6373-d194-4968-9783-62e9c818a617"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/b5c96365-9f56-42ed-ba61-f89d454ff395"> | 

</p>
</details> 

<details>
<summary>Custom Profile field (Date + Link)</summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/c07858ef-b1dc-45d3-bb11-863b6dd62032"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/568532e8-05ec-416e-9d65-847b7eaee0c7"> | 

</p>
</details> 

<details>
<summary>Custom Profile field (External Account + Pronouns)</summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/15b76800-c324-478f-8273-969f5c8852f7"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/1fd7e4b0-064a-4bc9-a3a2-62f9210d3897"> | 

</p>
</details> 

<details>
<summary>Long email and custom profile link</summary>
<p>

| Before | After |
|--------|--------|
| <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/d0bc5000-38ea-4ad6-88f7-cd05e1b90876"> | <img width="480" alt="image" src="https://github.com/zulip/zulip/assets/82862779/2c12a4a0-8725-4904-9224-424ae273200e"> | 

</p>
</details> 

---


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sayamsamal commented on web/styles/app_variables.css:
```diff
@@ -98,6 +98,7 @@
     --navbar-popover-menu-min-width: 230px;
     --message-actions-popover-min-width: 230px;
     --topic-actions-popover-min-width: 200px;
+    --user-card-popover-min-width: 200px;
```
I have started out with a min-width to be `200px` so that the names don't
look too compressed, but we would likely need to discuss this in CZO to
come with a min-width that has a better reasoning behind it.

alya commented:
This looks amazing!

alya commented:
Let's add a "Clear your status" tooltip to the user card, like we have in the menu.

![Screenshot 2024-07-02 at 12 43 05@2x](https://github.com/zulip/zulip/assets/2090066/f3795bf5-9b57-414c-a864-7b9d05591076)

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -161,33 +147,44 @@
-            background-color: transparent !important;
-        }
+    ~ i {
+        margin-left: auto;
```
We use `margin-left: auto` on the copy icon instead of `flex-grow: 1` on
the link item, since we don't want the underline to span across the whole
width of the element.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -1074,6 +1035,85 @@ ul {
     }
 }
 
+.popover-menu-user-header {
```
Created new class names instead of using the ones previously being used in
the personal settings popover with the intention of eventually replacing
the classes used in the personal settings popover with the CSS here.

timabbott replied:
Sounds good, but maybe we should spell it `popover-menu-user-card-header`

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -1074,6 +1035,85 @@ ul {
+    }
+
+    &.email_copied {
+        color: hsl(170deg 48% 54%) !important;
```
I have kept the color change on the email being copied intact, but this
color is standout from the rest of the UI, so we should discuss about it at CZO.

alya replied:
Ooh, yeah, I don't think we want this. I think we should just use a "Copied!" tooltip for consistency with the copying confirmations across the UI.

alya replied:
Let's check with @terpimost about the icon hover color as well (we can discuss in #design). I haven't checked the values, but it looks the same as here?

![Screenshot 2024-07-02 at 13 00 13](https://github.com/zulip/zulip/assets/2090066/6d641428-8019-45c3-8529-715c718905cb)

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -1251,6 +1291,8 @@ ul {
 
     .status_emoji {
+        align-self: flex-start;
+        flex-shrink: 0;
```
This prevents the status emoji from shrinking when we have a long status message.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -1284,15 +1326,27 @@ ul {
 }
 
+#user_card_popover {
+    --popover-menu-max-width: 250px;
```
I have also set a max-width of `250px` to prevent the custom profile fields
from extending too far, this also might need some dicussion.

timabbott replied:
That sounds prudent.

alya commented:
Testing keyboard navigation, I don't see a way to get to the "copy" button -- it gets skipped by arrows and Tab.

Also, super nitpicky (and definitely not a blocker), but I guess we could have better alignment here (this is 16px):

![Screenshot 2024-07-02 at 13 02 02](https://github.com/zulip/zulip/assets/2090066/5abd098c-bfe0-4141-a75c-b57600b2724a)

sayamsamal commented:
> Testing keyboard navigation, I don't see a way to get to the "copy" button -- it gets skipped by arrows and Tab.
> 
> Also, super nitpicky (and definitely not a blocker), but I guess we could have better alignment here (this is 16px):
> 
> ![Screenshot 2024-07-02 at 13 02 02](https://private-user-images.githubusercontent.com/2090066/345184840-5abd098c-bfe0-4141-a75c-b57600b2724a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk5NTU4NTUsIm5iZiI6MTcxOTk1NTU1NSwicGF0aCI6Ii8yMDkwMDY2LzM0NTE4NDg0MC01YWJkMDk4Yy1iZmUwLTQxNDEtYTc1Yy1iNTc2MDBiMjcyNGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDJUMjEyNTU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmZmMWNhNjc3Y2M2ZDgzY2ZlZjRjNmI0ZWNjNDk5ZWI4YTdkY2I3NDgyNDE3OTY5NjM0ODliYmZmOTNhYzhjNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.-uNoUC5SG_WL8v78HCJD9s5RpiQ8Vlq62jRRBY1bTEQ)

Is that blue outline, while being in focused? Since I seem to be getting that properly aligned.
<img width="284" alt="image" src="https://github.com/zulip/zulip/assets/82862779/a95ab95f-5fe6-4f46-9b3b-cf2d76581311">

alya commented:
Yeah, that's the focus outline I'm seeing (Chrome on a Mac; I don't recall if this was at 14px of 16px).

sayamsamal commented:
@amanagr could you please review this PR, apart from the feedback already mentioned? Thanks!

sayamsamal commented:
> Yeah, that's the focus outline I'm seeing (Chrome on a Mac; I don't recall if this was at 14px of 16px).

Weird, I tested it on Chrome on a Mac, and for both 14/122 and 16/140 but the focus outline looks aligned to me. 🤔

amanagr commented:
PR is kinda hard to review but I read through all the changes and they looked in the correct direction to me. I tested for different user states as in the PR description. I don't know how to test for unknown user @sayamsamal can you help me with that?

Other than that, looks right to me.

alya commented:
Unknown users are shown to limited guests when they see a message from someone who's not in their channels. https://zulip.com/help/guest-users

> For example, messages and reactions from a former subscriber of a channel will be shown as from an Unknown user.

sayamsamal commented:
@amanagr Thank you for the review, I appreciate the efforts knowing the complexity of this one. And as much as I tried to avoid it, the structural change of the popover following #30252, lead to some functional changes also being added to the design changes.

Also, I think @alya answers the part about how to test for a unknown user popover, but I'd like to add that a quick way to test it out without meeting all the conditions manually is to just unsubscribe the guest from all channels, and when you log back in as the guest, you'll mostly find some users in the left sidebar being now displayed as "Unknown user".

alya commented:
@sayamsamal Can this be reviewed for integration? I'm fine with keyboard nav being fixed up as a follow-up if that's helpful.

sayamsamal commented:
I have fixed the keyboard nav issue, and added some small improvements, pushing those in some time. After those, we can move ahead with the integration review.

sayamsamal commented:
## Updates:
- On copying the user's email, now a "Copied!" tooltip is shown, instead of replacing the email with "Email copied!" text.
- Added the Zulip color pallete from https://www.figma.com/design/msWyAJ8cnMHgOMPxi7BUvA/Zulip-Web-UI-kit?node-id=900-1861&t=EOMCSEzuY8AHg9EQ-0
- Changed the copy button hover and active state color using the new color palette and added background to the copy icons.
- Fixed alignment of the user info with respect to the user avatar.
- Cleanup of some unused code.
- Added missing tooltip from the clear status button.
- Fixed keyboard navigation for the copy buttons (Email + Custom profile field links)
- Improved the outline styling for the focused elements.

sayamsamal commented:
### Clear status button popover
<img width="646" alt="image" src="https://github.com/zulip/zulip/assets/82862779/6cdb736f-d059-4b58-9a9b-42776263ae00">

### Copy button on hover
<img width="222" alt="image" src="https://github.com/zulip/zulip/assets/82862779/44ab4e1e-9d3f-477e-a745-30648b9ac683">
<img width="222" alt="image" src="https://github.com/zulip/zulip/assets/82862779/3c9e3ddd-1877-40d0-b9c7-b0ec2c1ffc23">

### Copy button on active
<img width="222" alt="image" src="https://github.com/zulip/zulip/assets/82862779/72f3b080-1ce7-4b6b-9424-13a4400d1175">
<img width="222" alt="image" src="https://github.com/zulip/zulip/assets/82862779/f97f8d2f-9033-4b54-a6cf-3234f7ce29a2">

### Alignment with short user name vs long user name
<img width="354" alt="image" src="https://github.com/zulip/zulip/assets/82862779/03c8df99-b510-4149-82fa-f9aaa94f7f81">
<img width="354" alt="image" src="https://github.com/zulip/zulip/assets/82862779/be403a8b-009f-4c67-8bbe-5dc9152128ab">

### Keyboard navigation + outline styling in popovers

https://github.com/zulip/zulip/assets/82862779/05c64586-2162-4d20-8c3a-c90b6e3ed7b0

sayamsamal commented:
@alya I have pushed the fixes and improvements. This is ready for integration review, once you take a look at the new changes.

alya commented:
Thanks!

Hm, the "copy email" button isn't working for me for some reason. The email doesn't end up getting copied, and I don't get the "Copied!" confirmation tooltip.

alya commented:
I started a [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/copy.20button.20colors/near/1882639) about the "copy" button colors.

alya commented:
Everything else looks good to me in manual testing. If we want to tweak other things, we could do it as follow-ups.

sayamsamal commented:
> Thanks!
> 
> Hm, the "copy email" button isn't working for me for some reason. The email doesn't end up getting copied, and I don't get the "Copied!" confirmation tooltip.

The copy email button wasn't being hidden properly when there was no overflow. SInce we only show the email copy button when there is overflow, the logic for copying the mail via the button only works then as well.

alya commented:
Got it, makes sense.

alya commented:
@timabbott I think you can take a look now; we can figure out the colors in parallel.

sayamsamal commented:
Update: Removed the merge conflicts resulting from #30722.

timabbott commented on web/styles/popovers.css:
```diff
@@ -161,33 +161,44 @@
 }
 
-.user-card-popover-actions .custom_user_field {
+.custom-profile-field-value {
```
These selectors should probably still be scoped inside something about popovers; these presentation styles are not applicable to every custom profile field value we might display anywhere in the app.

timabbott commented on web/styles/popovers.css:
```diff
@@ -1054,6 +1034,89 @@ ul.popover-group-menu-member-list {
+    }
+}
+
+.user-email-field {
```
Again, this class should really have `user-card-popover` in its name if it's a top-level rule.

timabbott commented on web/templates/popovers/user_card/user_card_popover.hbs:
```diff
@@ -1,176 +1,207 @@
+                {{/if}}
+            </a>
+        </li>
+        <li role="none" class="popover-menu-list-item link-item">
```
This should certainly be a follow-up, possibly after 9.0, but `link-item` should probably be `popover-link-item`.

timabbott commented on web/styles/app_variables.css:
```diff
@@ -1,4 +1,167 @@
 :root {
+    /* Zulip color palette */
+    /* stylelint-disable color-no-hex */
```
I thought the color palette was defined using a fancier color space; @terpimost can you clarify what we want to write here? 

(I think we can probably fix in a follow-up, given the limited use of these in this PR).

timabbott commented on web/src/tippyjs.ts:
```diff
@@ -601,6 +601,12 @@ export function initialize(): void {
     });
 
+    tippy.delegate("body", {
+        target: ".personal-menu-clear-status",
```
It's a bit confusing to have this selector name be used for two different components; probably ideally we'd just use a different CSS class name for the one in the user card, and list them both here.

timabbott commented:
I read the code pretty carefully and identified three bugs:
- Initial keyboard focus wasn't set at all when using the keyboard shortcut `u` to open the user card popover for a sender. I debugged this and added a couple commits to fix it at the end of the branch. Given how this failed, I'm honestly not sure how this was working in `main` previously, so there may be more to investigate here as a follow-up -- we might have had two duplicate mechanisms here. @amanagr FYI as `user_card_popover: Fix on_mount hooks not being called` fixes what appears to be a regression in your code.
- Initial keyboard focus is not being set when clicking on the right sidebar to open the menu. This is less important, but it suggests to me that the actual fix is something like (1) we had some other more general mechanism for doing initial keyboard in popovers, and the `on_mount` hook that I fixed should actually be deleted in favor of that, (2) this mechanism was broken by refactoring in the same way that I fixed in `user_card_popover: Skip copy icon for keyboard focus.`. 
- The medium avatar code removed in https://github.com/zulip/zulip/pull/30663#discussion_r1674560428 was I think important.

I posted a handful of comments about code style things, all of which I think can be addressed in follow-ups if necessary. 

For the visuals:
- I'm not sure whether the fainter color labels are too faint. I think we can discuss in #design and fix-forward, since it's a fairly localized detail.
- I'm not sure whether the on-hover background color is too low-contrast; it feels hard to passively perceive on my external monitor. Probably we should darken that slightly.

sayamsamal commented:
> * Initial keyboard focus is not being set when clicking on the right sidebar to open the menu. This is less important, but it suggests to me that the actual fix is something like (1) we had some other more general mechanism for doing initial keyboard in popovers, and the `on_mount` hook that I fixed should actually be deleted in favor of that, (2) this mechanism was broken by refactoring in the same way that I fixed in `user_card_popover: Skip copy icon for keyboard focus.`.

This member function of the `PopoverMenu` class seems to be responsible for the keyboard handling of the user card popovers. The problem over here is the use of only the `a` tag as the selector, setting this to `a:visible` solves this issue. This can be investigated as such: Open your own user card from the popover, since here no `a` tag is hidden, the keyboard nav works fine. Now, the same for other users' user card popover doesn't work as some of the options are hidden there.

```js
handle_keyboard(key) {
    if (!this.is_open()) {
        blueslip.error("Trying to get the items when popover is closed.");
        return;
    }

    const $popover = $(this.instance?.popper);
    if (!$popover) {
        blueslip.error("Cannot find popover data.");
        return;
    }

    const $items = $("li:not(.divider):visible a", $popover);

    popover_items_handle_keyboard_with_overrides(key, $items);
}
```

timabbott commented:
OK cool, that looks great. One outstanding question is whether we can delete the `onMount` hook.

timabbott commented:
We've still got some fiddling with colors design discussions in flight, but I think we're at the point where the overall change is going to work; I'm going to merge this, but we should work hard to get those colors to a place we're excited about next week before 9.0 final. Thanks for all the work on this @sayamsamal!
